### PR TITLE
Using HfArgumentParser argument handling in the DPR engine

### LINKED ISSUE #412: DPR engine with HfArgumentParser argument handling
Convert the DPR engine to HfArgumentParser argument handling, as needed for integration with pipelines.

### PR DESCRIPTION
**Pull request by franzmpub**
# Pull Request      

## What does this PR do?

```Closes #412, #391```       

## Description

DPR engine with HfArgumentParser argument handling, including updates in code, CI tests, README.md and Jupyter notebook.
Adds saving the tokenizer trained model together with the trained model to allow uploading the complete model files to the HF repository (and using them by referencing the model name).
Renames arguments for better consistency between DPR training/indexing/search, and also usage in the ColBERT engine.
Fixes the `sample_negative_from_top_k` default value bug (Issue #391).
Removes unused arguments (e.g. `resume_from`).

Note: there are training/indexing/search, parameters not exposed in `primeqa/ir/dense/dpr_top/dpr/config.py`- the current version is a compromise allowing to change the parameters typically adjusted in experiments.
 
### Request Review

Be sure to **request a review** from one or more reviewers (unless the PR is to an unprotected branch).

## Versioning

When opening a PR to make changes to PrimeQA (i.e. `primeqa/`) master, be sure to increment the version following
[semantic versioning](https://semver.org/).  The VERSION is stored [here](https://github.com/primeqa/primeqa/blob/main/VERSION) 
and is incremented using `bump2version {patch,minor,major}` as described in the (development guide documentation)[development.html].

 - [x] Have you updated the VERSION?
 - [ ] Or does this PR not change the `primeqa` package or was not into master?

After pulling in changes from master to an existing PR, ensure the VERSION is updated appropriately.
This may require bumping the version again if it has been previously bumped.

If you're not quite ready yet to post a PR for review, feel free to open a [draft PR](https://github.blog/2019-02-14-introducing-draft-pull-requests/).

### After Merging
If merging into master and VERSION was updated, after this PR is merged:
- [ ] [Create a release](https://docs.github.com/en/github/administering-a-repository/releasing-projects-on-github/managing-releases-in-a-repository#creating-a-release) from the master with version equal to [VERSION](https://github.com/primeqa/primeqa/blob/main/VERSION)
- [ ] Not merging to master or VERSION not updated

### Checklist

Review the following and mark as completed:
- [x] [Tag an issue or issues](#what-does-this-pr-do) this PR addresses.
- [x] Added [description](#description) of changes proposed.
- [x] Review requested as [appropriate](#request-review).
- [x] Version bumped as [appropriate](#Versioning).
- [x] New classes, methods, and functions documented.
- [x] Documentation for modified code is updated.
- [ ] Built documentation to confirm it renders as expected (see [here](development.md)).
- [x] Code cleaned up and commented out code removed.
- [ ] Tests added to ensure all functionalities tested at >= 60% unit test coverage (see [here](development.md)).
- [ ] Code cleaned up and commented out code removed.
- [ ] Release created [as needed](#after-merging) after merging.

### CODE REVIEW

**Review by jsmcibm — APPROVED** [review]
LGTM after off-line discussion about choice of names, default value issues

File: primeqa/ir/README.md [hunk 1]
```diff
@@ -150,12 +150,12 @@ Using a model trained as described [here](https://github.com/primeqa/primeqa/tre
 python primeqa/ir/run_ir.py \
     --engine_type DPR \
     --do_index \
-    --dpr_ctx_encoder_path <context_encoder_model> \
+    --ctx_encoder_name_or_path <context_encoder_model> \
```
Comment (avisil) [inline primeqa/ir/README.md:153]: Can we please make this arg be `model_name_or_path` to be consistent?

File: primeqa/ir/README.md [hunk 2]
```diff
@@ -150,12 +150,12 @@ Using a model trained as described [here](https://github.com/primeqa/primeqa/tre
...
     --embed <part_number>of<parts_total> \
     --sharded_index \
```
Comment (avisil) [inline primeqa/ir/README.md:155]: Can this be left default?

File: primeqa/ir/README.md [hunk 3]
```diff
@@ -150,12 +150,12 @@ Using a model trained as described [here](https://github.com/primeqa/primeqa/tre
 python primeqa/ir/run_ir.py \
     --engine_type DPR \
     --do_index \
-    --dpr_ctx_encoder_path <context_encoder_model> \
+    --ctx_encoder_name_or_path <context_encoder_model> \
     --embed <part_number>of<parts_total> \
```
Comment (avisil) [inline primeqa/ir/README.md:154]: This arg too is very DPR specific and does it need to be exposed?

File: primeqa/ir/dense/dpr_top/dpr/biencoder_gcp.py [hunk 4]
```diff
@@ -28,7 +28,7 @@ class BiEncoder(torch.nn.Module):
     def __init__(self, hypers: BiEncoderHypers):
         super().__init__()
         self.hypers = hypers
-        logger.info(f'BiEncoder: initializing from {hypers.qry_encoder_name_or_path} and {hypers.qry_encoder_name_or_path}')
+        logger.info(f'BiEncoder: initializing from {hypers.qry_encoder_name_or_path} and {hypers.ctx_encoder_name_or_path}')
```
Comment (avisil) [inline primeqa/ir/dense/dpr_top/dpr/biencoder_gcp.py:31]: `model_name_or_path`

File: primeqa/ir/dense/dpr_top/dpr/dpr_util.py [hunk 5]
```diff
@@ -9,17 +9,11 @@
 class DPROptions(HypersBase):
     def __init__(self):
         super().__init__()
-        self.corpus_endpoint = ''
         self.port = 5001
-        self.model_name_or_path = ''  # used as qry_encoder_path
         self.rag_model_path = ''
```
Comment (avisil) [inline primeqa/ir/dense/dpr_top/dpr/dpr_util.py:13]: Are we running RAG ? I thought we just do DPR..

**Review by avisil — COMMENTED** [review]
Added some comments. Once address I'll take a look again. Thanks.

File: primeqa/ir/README.md (hunk 1 above)
Comment (franzmpub) [reply primeqa/ir/README.md:153]: Left alone as after a separate discussion.

File: primeqa/ir/README.md (hunk 3 above)
Comment (franzmpub) [reply primeqa/ir/README.md:154]: Left alone as after a separate discussion.

File: primeqa/ir/README.md (hunk 2 above)
Comment (franzmpub) [reply primeqa/ir/README.md:155]: The point here is to show how to use a model trained in the previous (training) step.

File: primeqa/ir/dense/dpr_top/dpr/biencoder_gcp.py (hunk 4 above)
Comment (franzmpub) [reply primeqa/ir/dense/dpr_top/dpr/biencoder_gcp.py:31]: Left alone as after a separate discussion.

File: primeqa/ir/dense/dpr_top/dpr/dpr_util.py (hunk 5 above)
Comment (franzmpub) [reply primeqa/ir/dense/dpr_top/dpr/dpr_util.py]: Good point, removed `self.rag_model_path = ''`

File: tests/primeqa/ir/dense/dpr_top/test_dpr_engine.py [hunk 6]
```diff
@@ -54,14 +59,16 @@ def test_engine(self, test_files_location):
             "--num_train_epochs", "2",
```
[the remark this answers is not in the document]
Comment (franzmpub) [reply tests/primeqa/ir/dense/dpr_top/test_dpr_engine.py]: Changed `--num_train_epochs` to `--epochs` as in the ColBERT model,  updated in code, CI test, README, Jupyter notebook

**Review by avisil — APPROVED** [review]
LGTM

File: primeqa/ir/dense/dpr_top/dpr/config.py [hunk 7]
```diff
@@ -0,0 +1,195 @@
+from dataclasses import dataclass, field
+
+
+@dataclass
+class DPRTrainingArguments:
```
Comment (kpfadnis) [inline primeqa/ir/dense/dpr_top/dpr/config.py:5]: For now, I don't see any issues but in future we might have use a standard config defining practice across PrimeQA and all utility methods parsing config should be part of that. 


Also we have to define a common mechanism to set something as a `choices` or `options` in metadata for a field.

**Review by kpfadnis — APPROVED** [review]
LGTM. 


In long term, we need to figure out standardized way of defining config parameters.

**Pull request merged**